dist/functional-test-k8s.sh

```diff
@@ -0,0 +1,117 @@
+EOF
+do
+    sleep 1
+done
+
+cat <<EOF |  docker run -i --rm --net=host gcr.io/google_containers/hyperkube-amd64:v$K8S_VERSION /hyperkube kubectl create -f - >/dev/null 2>/dev/null
+apiVersion: v1
+kind: Node
+metadata:
+  name: flannel2
+  annotations:
+    dummy: value
+spec:
+  podCIDR: 10.10.2.0/24
+EOF
+}
+
+teardown_suite() {
+    # Teardown the etcd server
+    docker rm -f flannel-e2e-test-etcd >/dev/null
+    docker rm -f flannel-e2e-k8s-apiserver >/dev/null
+}
+
+teardown() {
+	docker rm -f flannel-e2e-test-flannel1 >/dev/null 2>/dev/null
+	docker rm -f flannel-e2e-test-flannel2 >/dev/null 2>/dev/null
+}
+
+start_flannel() {
+    local backend=$1
+
+	flannel_conf="{ \"Network\": \"$FLANNEL_NET\", \"Backend\": { \"Type\": \"${backend}\" } }"
+    for host_num in 1 2; do
+       docker rm -f flannel-e2e-test-flannel$host_num >/dev/null 2>/dev/null
+       docker run -e NODE_NAME=flannel$host_num --privileged --name flannel-e2e-test-flannel$host_num -tid --entrypoint /bin/sh $FLANNEL_DOCKER_IMAGE >/dev/null
+       docker exec flannel-e2e-test-flannel$host_num /bin/sh -c 'mkdir -p /etc/kube-flannel'
+       echo $flannel_conf | docker exec -i flannel-e2e-test-flannel$host_num /bin/sh -c 'cat > /etc/kube-flannel/net-conf.json'
+       docker exec -d flannel-e2e-test-flannel$host_num /opt/bin/flanneld --kube-subnet-mgr --kube-api-url $k8s_endpt
+    done
+}
+
+create_ping_dest() {
+    # add a dummy interface with $FLANNEL_SUBNET so we have a known working IP to ping
+    for host_num in 1 2; do
+       # Use declare to allow the host_num variable to be part of the ping_dest variable name. -g is needed to make it global
+       declare -g ping_dest$host_num=$(docker "exec" --privileged flannel-e2e-test-flannel$host_num /bin/sh -c '\
+		source /run/flannel/subnet.env && \
+		ip link add name dummy0 type dummy && \
+		ip addr add $FLANNEL_SUBNET dev dummy0 && ip link set dummy0 up && \
+		echo $FLANNEL_SUBNET | cut -f 1 -d "/" ')
+    done
+}
+
+test_vxlan() {
+    start_flannel vxlan && sleep 2
+    create_ping_dest # creates ping_dest1 and ping_dest2 variables
+    pings
+}
+
+test_udp() {
+    start_flannel udp && sleep 2
+    create_ping_dest # creates ping_dest1 and ping_dest2 variables
+    pings
+}
+
+test_host-gw() {
+    start_flannel host-gw && sleep 2
+    create_ping_dest # creates ping_dest1 and ping_dest2 variables
+    pings
+}
+
+pings() {
+    # ping in both directions
+	assert "docker exec -it --privileged flannel-e2e-test-flannel1 /bin/ping -c 5 $ping_dest2" "Host 1 cannot ping host 2"
+	assert "docker exec -it --privileged flannel-e2e-test-flannel2 /bin/ping -c 5 $ping_dest1" "Host 2 cannot ping host 1"
+}
+
+test_manifest() {
+    # This just tests that the API server accepts the manifest, not that it actually acts on it correctly.
+    assert "cat ../Documentation/kube-flannel.yml |  docker run -i --rm --net=host gcr.io/google_containers/hyperkube-amd64:v$K8S_VERSION /hyperkube kubectl create -f -"
```

Comment:
btw, there's a cool project `garethr/kubeval` that can validate the k8s manifests. Not suggesting we should or shouldn't use it, but just throwing it out there